src/django_components/util/command.py

```diff
@@ -0,0 +1,407 @@
+    class. `ComponentCommand` allows you to define:
+
+    - Command name, description, and help text
+    - Arguments and options (e.g. `--name John`)
+    - Group arguments (see [argparse groups](https://docs.python.org/3/library/argparse.html#argument-groups))
+    - Subcommands (e.g. `components ext run my_ext hello`)
+    - Handler behavior
+
+    Each extension can add its own commands, which will be available to run with `components ext run`.
+
+    Extensions use the `ComponentCommand` class to define their commands.
+
+    For example, if you define and install the following extension:
+
+    ```python
+    from django_components ComponentCommand, ComponentsExtension
+
+    class HelloCommand(ComponentCommand):
+        name = "hello"
+        help = "Say hello"
+        def handle(self, *args, **kwargs):
+            print("Hello, world!")
+
+    class MyExt(ComponentsExtension):
+        name = "my_ext"
+        commands = [HelloCommand]
+    ```
+
+    You can run the `hello` command with:
+
+    ```bash
+    python manage.py components ext run my_ext hello
+    ```
+
+    You can also define arguments for the command, which will be passed to the command's `handle` method.
+
+    ```python
+    from django_components import CommandArg, ComponentCommand, ComponentsExtension
+
+    class HelloCommand(ComponentCommand):
+        name = "hello"
+        help = "Say hello"
+        arguments = [
+            CommandArg(name="name", help="The name to say hello to"),
+            CommandArg(name=["--shout", "-s"], action="store_true"),
+        ]
+
+        def handle(self, name: str, *args, **kwargs):
+            shout = kwargs.get("shout", False)
+            msg = f"Hello, {name}!"
+            if shout:
+                msg = msg.upper()
+            print(msg)
+    ```
+
+    You can run the command with:
+
+    ```bash
+    python manage.py components ext run my_ext hello --name John --shout
+    ```
+
+    !!! note
+
+        Command arguments and options are based on Python's `argparse` module.
+
+        For more information, see the [argparse documentation](https://docs.python.org/3/library/argparse.html).
+    """
+
+    name: str
+    """The name of the command - this is what is used to call the command"""
+    help: Optional[str] = None
+    """The help text for the command"""
+    arguments: Sequence[Union[CommandArg, CommandArgGroup]] = ()
+    """argparse arguments for the command"""
+    subcommands: Sequence[Type["ComponentCommand"]] = ()
+    """Subcommands for the command"""
+
+    handle: Optional[CommandHandler] = None
+    """
+    The function that is called when the command is run. If `None`, the command will
+    print the help message.
+    """
+
+    parser_input: Optional[CommandParserInput] = None
+    """
+    The input to use when creating the `ArgumentParser` for this command. If `None`,
+    the default values will be used.
+    """
+    subparser_input: Optional[CommandSubcommand] = None
```

Comment:
Note: ArgumentParser supports defining subcommands (they call it [subparsers](https://docs.python.org/3/library/argparse.html#sub-commands)). However, the fields that can be set for the subparser are slightly different than what can be set when initializing `ArgumentParser`.

Wasn't sure how to approach this, so for simplicity I kept the two separate:
- If a command class is being used as a top-level command, we're creating an instance of `ArgumentParser`, and in that case `setup_parser_from_command()` uses the settings from the `parser_input` field
- But if the command class is being used as a subcommand, we're not instantiating `ArgumentParser` directly, but instead call `ArgumentParser.add_subparsers().add_parser()`. And in this case we'll pass the settings in `subparser_input`.